File: q1026/q1026.java
class Solution {

    // track the max diff seen during DFS
    private int maxDiff = 0;

    public int maxAncestorDiff(TreeNode root) {
        dfsHelper(root, root.val, root.val);
        return maxDiff;
    }

    private void dfsHelper(TreeNode node, int maxVal, int minVal) {
        // update the max diff based on the node, maxVal, and minVal we have seen
        if (node == null)
            return;
        int diff1 = Math.abs(node.val - maxVal);
        int diff2 = Math.abs(node.val - minVal);
        maxDiff = Math.max(Math.max(diff1, diff2), maxDiff);

        // update the maxVal and minVal
        maxVal = Math.max(node.val, maxVal);
        minVal = Math.min(node.val, minVal);

        // recurse on the left and right subtree
        dfsHelper(node.left, maxVal, minVal);
        dfsHelper(node.right, maxVal, minVal);
    }
}

File: q1026/q1026.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:

        # track the max diff we see during the dfs
        self.maxDiff = 0

        def dfsHelper(node, maxPrev, minPrev): 
            if node == None: return
            currVal = node.val

            # max diff is either constructed by curr val - the max / min of it ancestor
            currDiff = max(abs(maxPrev - currVal), abs(minPrev - currVal))
            self.maxDiff = max(currDiff, self.maxDiff)

            # update the max and min of ancestor before recursing
            maxPrev = max(currVal, maxPrev)
            minPrev = min(currVal, minPrev)

            # recursively deal with the left and right subtree
            dfsHelper(node.left, maxPrev, minPrev)
            dfsHelper(node.right, maxPrev, minPrev)
        
        dfsHelper(root, root.val, root.val)
        return self.maxDiff
        